File: examples/ruby/lib/word_count.rb
# WordCount class that emulates the behavior of the Unix wc command
class WordCount
  attr_reader :file_path

  def initialize(file_path)
    @file_path = file_path
    @content = read_file
  end

  # Count the number of lines in the file
  # wc counts lines by counting newline characters
  def line_count
    @line_count ||= @content.scan(/\n/).length
  end

  # Count the number of words in the file
  def word_count
    @word_count ||= @content.split.count
  end

  # Count the number of characters in the file
  def character_count
    @character_count ||= @content.length
  end

  # Return a hash with all counts
  def count
    {
      lines: line_count,
      words: word_count,
      characters: character_count,
      file_path: file_path
    }
  end

  # Return formatted output matching wc command format
  def to_s
    sprintf("%8d %7d %7d %s", line_count, word_count, character_count, file_path)
  end

  private

  def read_file
    File.read(file_path)
  rescue Errno::ENOENT
    raise Errno::ENOENT, "No such file or directory - #{file_path}"
  end
end 

File: examples/elixir/lib/word_count.ex
defmodule WordCount do
  @moduledoc """
  A module that emulates the behavior of the Unix `wc` command.

  This module provides functions to count lines, words, and characters in a file,
  matching the exact behavior of the Unix `wc` command.
  """

  defstruct [:file_path, :content]

  @doc """
  Creates a new WordCount struct by reading the specified file.

  ## Parameters
  - file_path: String path to the file to analyze

  ## Returns
  - {:ok, %WordCount{}} on success
  - {:error, reason} on failure (e.g., file not found)

  ## Examples
      iex> {:ok, wc} = WordCount.new("../../data/file.txt")
      iex> wc.file_path
      "../../data/file.txt"
  """
  def new(file_path) do
    case File.read(file_path) do
      {:ok, content} ->
        {:ok, %WordCount{file_path: file_path, content: content}}

      {:error, reason} ->
        {:error, reason}
    end
  end

  @doc """
  Counts the number of lines in the file.

  Lines are counted by counting newline characters, matching wc behavior.

  ## Parameters
  - wc: %WordCount{} struct

  ## Returns
  - Integer count of lines
  """
  def line_count(%WordCount{content: content}) do
    content
    |> String.graphemes()
    |> Enum.count(&(&1 == "\n"))
  end

  @doc """
  Counts the number of words in the file.

  Words are counted by splitting on whitespace, matching wc behavior.

  ## Parameters
  - wc: %WordCount{} struct

  ## Returns
  - Integer count of words
  """
  def word_count(%WordCount{content: content}) do
    content
    |> String.split()
    |> length()
  end

  @doc """
  Counts the number of characters in the file.

  ## Parameters
  - wc: %WordCount{} struct

  ## Returns
  - Integer count of characters
  """
  def character_count(%WordCount{content: content}) do
    String.length(content)
  end

  @doc """
  Returns all counts as a map.

  ## Parameters
  - wc: %WordCount{} struct

  ## Returns
  - Map with keys :lines, :words, :characters, :file_path
  """
  def count(wc) do
    %{
      lines: line_count(wc),
      words: word_count(wc),
      characters: character_count(wc),
      file_path: wc.file_path
    }
  end

  @doc """
  Formats the output to match the wc command format.

  ## Parameters
  - wc: %WordCount{} struct

  ## Returns
  - String formatted like wc command output
  """
  def format_output(wc) do
    lines = line_count(wc)
    words = word_count(wc)
    chars = character_count(wc)

    :io_lib.format("~8w ~7w ~7w ~s", [lines, words, chars, wc.file_path])
    |> IO.iodata_to_binary()
  end
end
